### PR TITLE
Use latest protocompile in `buf format`, fix things up in the formatter

### PR DESCRIPTION
**Pull request by jhump**
There were a few bugs in the formatter that were uncovered with the new version of `protocompile`, mostly around trailing vs. leading comments.

For example, there was previously a bug in protocompile (https://github.com/bufbuild/protocompile/pull/33) where certain elements (like extension names enclosed in parentheses in an option name, full options declaration enclosed in brackets) would not get leading comments. The comments that _should_ be interpreted as leading would usually instead be interpreted as a trailing comment on the previous token.

That means that there were some cases in the formatter where it was not correctly handling leading comments on certain elements (such as custom option names).

Also, this was treating leading/trailing comments for inline comments in between tokens as if the parser's classification of trailing vs. leading were authoritative. If it was a trailing comment in the AST, it would format it with _no whitespace_ after the token, before the comment. Similarly, for leading comments, it would format it with nothing after the comment, before the token. But this classification of trailing vs. leading is just heuristic (and mostly done to mirror this distinction in source code info in descriptors generated by `protoc`). So now, the formatter will preserve whitespace: if a comment had no whitespace between it and a token, that will be preserved. If it did have whitespace, it will be canonicalized to a single space.

Getting everything looking right involved a greater degree of changes to the formatter than I would have liked. 🤷 

The good news is that these changes should mostly be idempotent for users: they shouldn't see a lot of new diff churn from formatting with this new version. As you'll see in the new golden files, the only new differences are the whitespace around inline comments. And this change tries to _preserve_ existing whitespace there, so if a file was previously formatted with the old version and then re-formatted with the new, there shouldn't be a diff.

### CODE REVIEW

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:117]: I added code to try to preserve some of the whitespace conventions in inline comments. That means it's possible for it to now call `f.Space()` but then the next call wants to write a newline. And trailing whitespace looks ugly.

So we now defer actually writing the space and skip it if the next character to write is a newline.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:144]: I removed the argument because it wasn't necessary since the file node is a field on `f`. In fact, `writeFileHeader` and `writeFileTypes` were also passed this argument but never even used it. So this is just minor clean up.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:433]: Previously, the option name rarely had a leading comment, which is why this bug was never noticed before.

We've only written all of the leading comments for compact options. Now that the new protocompile AST is likely to contain leading comments here, this bug was noticed (by seeing the formatter completely _drop_ a comment for a non-compact option declaration... oops!).

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:446]: Oops. This was also a place that was possibly dropping comments.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:1045]: This is kind of gross, to stash this state on `f` instead of on the stack. But weaving this flag all the way through numerous function calls, to get it into `writeOptionName` was even messier. So I went with the less smelly approach.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:1263]: Another bug. This code path was unexercised with the old version protocompile -- again because the comment would usually have been attributed as a trailing comment to previous token, so we'd not fall into this condition.

But with the new protocompile, it popped up as a mysterious missing newline in between elements (all because some logic that decides to add a newline wants to know if the last thing was a closing brace). Other call paths through this function do set this correctly after writing it.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:1299]: This is part of the logic that tries to preserve whitespace conventions. You'll see this condition sprinkled in a few places.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:1335]: No such thing as a bool literal in the new version of protocompile. We don't actually know if `true` or `false` is a boolean literal until we do type checking, when interpreting options. It could also be the name of an enum value. So now the AST will just use identifier nodes whose text might be `"true"` or `"false"`.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:1949]: It seemed more pleasing to the eye to always separate the comment with a space if it follows the end of a declaration (a semicolon or close brace).

File: private/buf/bufformat/formatter_test.go
Comment (jhump) [inline private/buf/bufformat/formatter_test.go:56]: Bug is fixed! https://github.com/bufbuild/protocompile/pull/34

File: private/buf/bufformat/formatter_test.go
Comment (jhump) [inline private/buf/bufformat/formatter_test.go:90]: Bug is fixed! https://github.com/bufbuild/protocompile/pull/33

File: private/buf/bufformat/formatter_test.go
Comment (jhump) [inline private/buf/bufformat/formatter_test.go:65]: Putting each file in its own test case provides much nicer output IMO. It also reduces the abruptness of a `require` assertion fail -- the test case that is immediately failed is just a single file instead of the entire function and all files.

File: private/buf/bufformat/formatter.go
Comment (jhump) [inline private/buf/bufformat/formatter.go:86]: This prevents doubled whitespace from accidentally popping up, now that there are more calls to `f.Space()`. In particular, we explicitly put a space between some tokens, but then we might also try to add a space when printing inline comments, and two spaces looks bad.

File: private/buf/bufformat/testdata/proto2/enum/v1/enum.golden.proto
Comment (jhump) [inline private/buf/bufformat/testdata/proto2/enum/v1/enum.golden.proto:8]: As can be seen below, the only diffs in the golden files are in whitespace around inline comments. If you look at the original files, you'll see we only expect a space in the golden output for cases where the original also had whitespace.

**Comment by lrewega** [comment]
> The good news is that these changes should mostly be idempotent for users: they shouldn't see a lot of new diff churn from formatting with this new version. As you'll see in the new golden files, the only new differences are the whitespace around inline comments. And this change tries to preserve existing whitespace there, so if a file was previously formatted with the old version and then re-formatted with the new, there shouldn't be a diff.

Do the golden files need to be altered at all then? I only did a cursory read of the patch, but I would expect to add new tests that demonstrate preservation of whitespace, making it clear that previous formatting is still valid.

**Comment by jhump** [comment]
> Do the golden files need to be altered at all then? 

@lrewega, Yes. The old logic did _not_ try to preserve whitespace. So the existing golden files are missing whitespace where the original source had it.

> I would expect to add new tests that demonstrate preservation of whitespace, 

The new golden files, compared to the existing originals, do demonstrate that.

> making it clear that previous formatting is still valid.

But to more specifically demonstrate that the previous formatting doesn't get churned, I just did manual testing. I pulled all of the files into a separate directory and run `buf format -w` and the golden files were not touched.

```
jhumphries@Joshuas-MacBook-Pro:/Users/jhumphries/src/tmp/test (master)> /Users/jhumphries/.cache/buf/Darwin/arm64/gobin/buf format -w
jhumphries@Joshuas-MacBook-Pro:/Users/jhumphries/src/tmp/test (master *)> git status | grep golden                                     
jhumphries@Joshuas-MacBook-Pro:/Users/jhumphries/src/tmp/test (master *)> git status              
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   customoptions/options.proto
	modified:   proto2/enum/v1/empty.proto
	modified:   proto2/enum/v1/enum.proto
	modified:   proto2/enum/v1/enum_trailing_comment.proto
	modified:   proto2/enum/v1/enum_trailing_newline.proto
	modified:   proto2/enum/v1/enum_value_trailing_comment.proto
	modified:   proto2/extend/v1/empty.proto
	modified:   proto2/field/v1/option.proto
	modified:   proto2/group/v1/empty.proto
	modified:   proto2/header/v1/import.proto
	modified:   proto2/header/v1/option.proto
	modified:   proto2/header/v1/package.proto
	modified:   proto2/license/v1/enum.proto
	modified:   proto2/license/v1/message.proto
	modified:   proto2/message/v1/empty.proto
	modified:   proto2/message/v1/message.proto
	modified:   proto2/message/v1/message_extensions.proto
	modified:   proto2/message/v1/message_group.proto
	modified:   proto2/message/v1/message_import.proto
	modified:   proto2/message/v1/message_options.proto
	modified:   proto2/message/v1/multiple_nested.proto
	modified:   proto2/message/v1/nested.proto
	modified:   proto2/message/v1/sparse.proto
	modified:   proto2/option/v1/option.proto
	modified:   proto2/option/v1/option_array.proto
	modified:   proto2/option/v1/option_complex_array_literal.proto
	modified:   proto2/option/v1/option_compound_name.proto
	modified:   proto2/option/v1/option_message_field.proto
	modified:   proto2/option/v1/option_name.proto
	modified:   proto2/option/v1/single_compact_option.proto
	modified:   proto3/all/v1/all.proto
	modified:   proto3/file/v1/option.proto
	modified:   proto3/header/v1/header.proto
	modified:   proto3/literal/v1/compound_string.proto
	modified:   proto3/literal/v1/literal.proto
	modified:   proto3/literal/v1/literal_comments.proto
	modified:   proto3/literal/v1/special_literal.proto
	modified:   proto3/oneof/v1/multiple.proto
	modified:   proto3/oneof/v1/simple.proto
	modified:   proto3/range/v1/range.proto
	modified:   proto3/service/v1/empty.proto
	modified:   proto3/service/v1/empty_rpc.proto
	modified:   proto3/service/v1/service.proto
	modified:   proto3/service/v1/service_options.proto
	modified:   proto3/service/v1/simple.proto

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	buf.yaml

no changes added to commit (use "git add" and/or "git commit -a")
```

**Review by joshcarp — APPROVED** [review]
Seems pretty straight forward to me, but then again this is the first part of `bufformat` that I've seen

**Review by bufdev — APPROVED** [review]
Really what I looked at was the golden tests, which make sense.

Add something to the changelog, i.e. "Fixed bug in formatter where spaces before inline comments were not always properly added."

**Comment by jhump** [comment]
Just pushed a minor commit based on feedback from @pkwarren: a "dangling" space before a semicolon looks funny. So we won't try to preserve that whitespace.

File: private/buf/bufformat/testdata/proto2/option/v1/option_message_field.golden.proto
Comment (pkwarren) [inline private/buf/bufformat/testdata/proto2/option/v1/option_message_field.golden.proto:54]: nit: This is another place where adding extra space after `[` and before `]` looks like it could be trimmed.

```suggestion
        [/* One */ foo.bar.Test.Nested._NestedNested._garblez /* Two */] /* Three */ : "boo" // Trailing comment on extension name.
```

File: private/buf/bufformat/testdata/proto2/option/v1/option_name.golden.proto
Comment (pkwarren) [inline private/buf/bufformat/testdata/proto2/option/v1/option_name.golden.proto:20]: nit: Looks nicer here if there aren't added spaces added after '(' and before ')'.

```suggestion
    (/* One */ another /* Two */) = 2
```

**Review by pkwarren — APPROVED** [review]
Very nice improvement (and good to see the new protocompile get pulled in with fixes for some previous bugs).

**Comment by bufdev** [comment]
Don't forget changelog!

File: private/buf/bufformat/testdata/proto2/message/v1/message_options.proto
Comment (jhump) [inline private/buf/bufformat/testdata/proto2/message/v1/message_options.proto:18]: @pkwarren, I updated the formatter like you requested, to omit whitespace after open parens and braces and before close parens and braces.

I then decided to overhaul this file with a bunch of new examples with various kinds of braces/brackets/parens, to make sure the output looked good. And I discovered a few issues.

Unfortunately, the fixes to these issues _will_ produce some diff churn for users .But I think it is worth it because the new canonical format is _much_ easier to read IMO.

Here are the things I fixed:
* If an array literal or message literal had an inline comment before the closing brace, it was printed in a rather verbose and bad-looking way:
   ```
   [
      123

   /* inline comment */
   ]
   ```
   So I've fixed this to look better IMO:
   ```
   [
      123
      /* inline comment */
   ]
   ```
* There was an exception for simple compact options to be printed inline, on a single line. But there was no such exception for simple (single element) arrays and messages. So I've added one. If the array or message has a single element/field whose value is not a nested array or element, it can be printed all on one line. You can see an example in field `d` of the complex option below.

**Comment by pkwarren** [comment]
Once the CI test failure is fixed and a changelog entry is added this looks good to me. Much more readable formatting.

File: private/buf/bufformat/testdata/proto2/message/v1/message_options.proto
Comment (jhump) [reply private/buf/bufformat/testdata/proto2/message/v1/message_options.proto:18]: Welp there are some nasty issues in this code. One is that it interprets "trailing comment" to mean something a little different than what it is. And, honestly, this probably warrants a change in protocompile.

Currently, the lexer attempts to classify comments as "trailing" for previous token or "leading" for current/next token. It does this using heuristics that try to emulate how `protoc` does this, for purposes of matching its source code info in descriptors.

_However_, those heuristics are too complicated and cause the interpretation of a comment as leading vs trailing to vary with subtle newline changes. That means that the act of formatting the code like above (removing an extra newline for compact options, message literals, and array literals) and removing a newline can cause a comment to be interpreted in a re-format as a trailing comment (when it was previously interpreted as a leading comment). That means that the format is inherently unstable: reformatting in that case would change it further.

I think what we really want is that the AST only calls something a trailing comment if it is on the same line as the element. And then, when computing source code info for the AST, the compiler may have to "borrow" some leading comments from the next token to populate trailing comments like `protoc` does. This "borrow" step being in the lexer and impacting the AST is the root issue right now.

Also, this code is very hard to maintain. There are are _lots_ of helper functions that are not named what they are ostensibly _for_ instead of being named after what they actually do. For example, there is `writeBodyEnd`, which is used to write the end of a body... but also used in myriad other places, such as elements in a compound string literal 🤷.

The way the helper methods are shared/re-used, but without clear names, means its easy to make a seemingly reasonable change in a helper but it to have surprising side effects because of how its re-used in weird places. Anyhow, I feel like I've only made the code messier in trying to get this to look a little better. And now I know just how brittle it is.

I think I will put this PR on hold and go and make the above-described changes to protocompile -- so the AST is more predictable and simple with what gets classified as a trailing comment and move the logic to classify like `protoc` does into its `sourceinfo` package. That will be a much more stable point on which to implement formatter improvements.

**Comment by jhump** [comment]
I think this will help: https://github.com/bufbuild/protocompile/pull/46

**Comment by jhump** [comment]
Okay. I think I've finally wrestled this thing into shape.

It's not pretty. I ended up making quite a lot of changes with the _hope_ of simplifying some of the nasty edge cases and whitespace handling. I'm not sure how well I succeeded (though the last commit is a net removal of about 80 LOC...).

But I think the golden files are looking good, including with some of the additions I made to test some add'l edge cases.

**Comment by jhump** [comment]
@pkwarren, @bufdev, I think this is finally ready to go. Though I probably need to double-check the differences in golden outputs and amend/improve the change log wording before merging... PTAL.

File: private/buf/bufformat/formatter.go
Comment (pkwarren) [inline private/buf/bufformat/formatter.go:1949]: Does this need to take into account indentation level?

File: private/buf/bufformat/formatter.go
Comment (pkwarren) [inline private/buf/bufformat/formatter.go:47]: It might be nice to add comments to these - there are some inline comments around use of defer that might make sense to move here?

File: private/buf/bufformat/testdata/proto2/message/v1/message_import.golden.proto
Comment (pkwarren) [inline private/buf/bufformat/testdata/proto2/message/v1/message_import.golden.proto:5]: This is one place where I feel like the original formatting reads a little nicer, however this does better to keep the original format of the file in place.

**Review by pkwarren — APPROVED** [review]
Really nice job with refactoring and fixing some edge cases. The golden files look much better to me and it is good that the bugs where comments were dropped is now fixed. The code in the formatter reads much nicer too.

Just a couple of minor comments (feel free to ignore the nit picks).

File: private/buf/bufformat/formatter.go
Comment (jhump) [reply private/buf/bufformat/formatter.go:1949]: It does. `f.P(...)` prints the indentation, then the given args, then a newline. (If not args, indentation is skipped and it just prints empty line.)

File: private/buf/bufformat/formatter.go
Comment (jhump) [reply private/buf/bufformat/formatter.go]: done

File: private/buf/bufformat/formatter.go
Comment (jhump) [reply private/buf/bufformat/formatter.go]: Doh! I'm glad you're looking closely. I've left this kind of crap in on several recent PRs 🤦 

Removed.

**Pull request merged**
